shell/components/xen/xen-state.js

```diff
@@ -49,58 +49,70 @@ export default Base => class extends Base {
     Object.assign(this._state, state);
     this._invalidate();
   }
+  _setIfDirty(object) {
+    let dirty = false;
+    const state = this._state;
+    for (const property in object) {
+      const value = object[property];
+      if (state[property] !== value) {
+        dirty = true;
+        state[property] = value;
+      }
+    }
+    if (dirty) {
+      this._invalidate();
+      return true;
+    }
+  }
   _async(fn) {
-    // TODO(sjmiles): SystemJS throws unless `Promise` is `window.Promise`
     return Promise.resolve().then(fn.bind(this));
     //return setTimeout(fn.bind(this), 10);
   }
   _invalidate() {
     if (!this._validator) {
-      //this._log('register _async validate');
-      //console.log(this.localName + (this.id ? '#' + this.id : '') + ': invalidated');
       this._validator = this._async(this._validate);
     }
   }
+  _getStateArgs() {
+    return [this._props, this._state, this._lastProps, this._lastState];
+  }
   _validate() {
+    const stateArgs = this._getStateArgs();
     // try..catch to ensure we nullify `validator` before return
     try {
-      // TODO(sjmiles): should this be a replace instead of a merge?
+      // TODO(sjmiles): should be a replace instead of a merge
       Object.assign(this._props, this._pendingProps);
       if (this._propsInvalid) {
         // TODO(sjmiles): should/can have different timing from rendering?
-        this._willReceiveProps(this._props, this._state, this._lastProps, this._lastState);
+        this._willReceiveProps(...stateArgs);
         this._propsInvalid = false;
       }
-      if (this._shouldUpdate(this._props, this._state, this._lastProps, this._lastState)) {
+      if (this._shouldUpdate(...stateArgs)) {
         // TODO(sjmiles): consider throttling render to rAF
         this._ensureMount();
-        this._update(this._props, this._state, this._lastProps, this._lastState);
+        this._update(...stateArgs);
       }
     } catch (x) {
       console.error(x);
     }
     // nullify validator _after_ methods so state changes don't reschedule validation
     // TODO(sjmiles): can/should there ever be state changes fom inside _update()? In React no, in Xen yes (until I have a good reason not too).
     this._validator = null;
+    // invalidation is live in _didUpdate
+    this._didUpdate(...stateArgs);
```

Comment:
Just double checking -- so previously we called `_didUpdate` below with different props/state than we will now, right? And that's ok because of changes to how invalidation is handled?

Rather than reading in depth to determine myself I figured better to just ask briefly, which means perhaps we should also explain here with a comment in the code. Or maybe that's what the comment at line 99 is getting at?

Reply:
It's ok to change this because `_didUpdate` is rarely used, and where it is used this change is irrelevant.

There is no good reason it should have a different state configuration then the methods that precede it, so I consider this a bug fix.

There's probably also no good reason for it to be outside of the validation sequence, but I wanted to limit the scope of changes.